### PR TITLE
Make current_version obey app & appversion param for langpacks, in add-on detail API

### LINKED ISSUE #8180: Do not mark an addon as incompatible if there is a previous version that is compatible.
Currently if you visit addons.mozilla.org and look at language packs, with Firefox 59, you'll see `This add-on is not compatible with your version of Firefox.`

Though in the "More Versions" link at the bottom left you will see a compatible version.

This is because language packs all set min_version and max_version to be very restrictive (only one major version of Firefox) rather than allowing a wider range. And because now Release Engineering submits the addon as listed ahead of the release day (previously AMO staff would manually take the language packs and submit them into AMO sometime shortly after the release day).

This has the end result of confusing users who may have an up to date Firefox (or even after the next version is out, if we stop updates due to crashes, or are throttling the rollout % of the new version) they won't be able to download the language pack they want, and are unlikely to look in "More Versions"

I propose that we offer them the most-recent addon version that is supported by the browser version they are loading AMO with, e.g. if they come in with Firefox 58 they get a direct link to download a language pack supported by Firefox 58, and similarly for 59, 60, etc.

This has the added bonus that we can realistically list language packs for each Firefox Beta version, and maybe even Nightly on AMO without worrying that we might be confusing users on the Release Version.

### PR DESCRIPTION
**Pull request by diox**
Only for langpacks and only in the add-on detail API, overwrite the `current_version` property we return by the latest public version compatible with what is passed in the app and appversion parameters, if both are present.

Fix #8180 (at least the API side of things)

### CODE REVIEW

File: src/olympia/addons/serializers.py [hunk 1]
```diff
@@ -211,6 +212,48 @@ class Meta:
                   'release_notes', 'reviewed', 'url', 'version')
 
 
+class CurrentVersionSerializer(SimpleVersionSerializer):
+    def to_representation(self, obj):
+        # If the add-on is a langpack, and `appversion` is passed, try to
+        # determine the latest public compatible version and replace the obj
+        # with the result. Because of the perf impact, only done for langpacks
+        # in the detail API.
```
Comment (eviljeff) [inline src/olympia/addons/serializers.py]: We're not doing anything to lesson the performance impact though, are we?  I don't see any extra caching, etc.  Is anything planned in a follow-up, etc?  Or are we strictly just limiting it by it only being langpacks?

File: src/olympia/addons/serializers.py [hunk 2]
```diff
@@ -211,6 +212,48 @@ class Meta:
...
+        passed through the request, or fall back to addon.current_version if
+        none is found.
+
+        Only use on langpacks if the appversion parameter is present.
+        """
+        request = self.context.get('request')
+        try:
+            application = AddonAppQueryParam(request).get_value()
+        except ValueError:
+            raise exceptions.ParseError('Invalid or missing app parameter.')
```
Comment (eviljeff) [inline src/olympia/addons/serializers.py]: I know this is how we've done it in the other places in view.py that uses `AddonAppQueryParam` but, imo, this logic should be defined once in the filter itself, like the other filters do.

File: src/olympia/addons/serializers.py [hunk 3]
```diff
@@ -211,6 +212,48 @@ class Meta:
...
+        try:
+            # AddonAppVersionQueryParam.get_values() returns (app_id, min, max)
+            # but we want {'min': min, 'max': max}.
+            appversions = dict(
+                zip(('min', 'max'),
+                    AddonAppVersionQueryParam(request).get_values()[1:]))
+        except ValueError:
+            raise exceptions.ParseError('Invalid appversion parameter.')
```
Comment (eviljeff) [inline src/olympia/addons/serializers.py]: Actually, why isn't `AddonAppVersionQueryParam` used?  It already has validation for app and appversion and raises an appropriate error message.

File: src/olympia/addons/tests/test_serializers.py [hunk 4]
```diff
@@ -667,13 +668,166 @@ class TestAddonSerializerOutput(AddonSerializerOutputTestMixin, TestCase):
     serializer_class = AddonSerializer
     serializer_class_with_unlisted_data = AddonSerializerWithUnlistedData
 
+    def setUp(self):
+        super(TestAddonSerializerOutput, self).setUp()
+        self.action = 'detail'
```
Comment (eviljeff) [inline src/olympia/addons/tests/test_serializers.py]: Isn't it 'retrieve'?

File: src/olympia/addons/serializers.py (hunk 1 above)
Comment (diox) [reply src/olympia/addons/serializers.py]: Stricly limiting it by it only being langpacks

File: src/olympia/addons/serializers.py (hunk 3 above)
Comment (diox) [reply src/olympia/addons/serializers.py]: Probably right, I'll check & remove the `AddonAppQueryParam` usage.

File: src/olympia/addons/tests/test_serializers.py (hunk 4 above)
Comment (diox) [reply src/olympia/addons/tests/test_serializers.py]: It is, good catch. It doesn't matter anywhere at the moment but I'll fix it (the tests for the langpack version do use `retrieve` explicitly.)

**Pull request merged**